### ISSUE TITLE
Allow monkey patch builtin functions

### ISSUE THREAD

**Issue by rominf** [fish-shell/fish-shell#1652]
#### Situation 1
I often try some things to understand how fish works, test my idea and so on. The first time I did it I wrote a function called `test` (unintentionally). Can you guess what happen next? Yes, this shell became unusable instantly after redefining the standard `test` function.

#### Situation 2
@terlar is writing testing framework (https://github.com/terlar/fish-tank). It has functions like `assert`, `assert_equal`, `assert_match`, etc.

What's wrong with it? There are many things that we could want to check (patterns, file existance, ...). Do we need to have tens of `assert`s?

Wouldn't be good to use standard functions (`test`, `contains`, ...) and programs (`expr`, `sed`) to do all the checking? For example, string equivalence can be asserted like this:
```
assert [ 42 = (find 'the meaning of life, the universe and everything' | basename (pwd)) ]
```

It would, but unfortunately, in this case testing framework wouldn't have required information about the input (like actual and expected results).

#### So
[Beware! I'm not familiar with the fish internals!]

I propose to make all builtin functions (like `test`) and programs non overwritable (meaning you cannot change the internals of shell simply by overwriting `test` function or `sed` by creating function with the same name).

How to achieve this? I will take functions for simplicity. First of all, all builtin functions should be underscored (`test` -> `_test`) and only underscored variants should be used for the internal needs. But, to provide the user the ability to use them normally there should be wrapping functions:
```
function test
  _test $argv
end
```
All builtin underscored functions should be protected for shadowing (defining function `_test` should produce an error).

If the user wants to change the function `test` he simply defines it and it doesn't break the shell! He also can use the original function by executing `_test`.

To sum up, both situation would be fixed without complicated architectural changes.

PS: to prevent programs overwriting it's possible to use something like `(which sed)` everywhere where you need `sed`.

**Comment by faho** [comment]
I don't believe there's anything left to do here since #3013.
